### DISCUSSION TITLE
Backward masking in jsPsych?

### DISCUSSION THREAD

**Discussion by spiropan** [jspsych/jsPsych#238] [Q&A]
Hello,
I would like to be able to backward mask an image in jsPsych (i.e. display a positive valence image for 33 ms immediately followed by a neutral mask for 167 ms). The easiest plugin to do this appears to be jspsych-animation.js. However, the frame_time parameter only accepts a single scalar value so the image and the mask would have to have the same frame rate. Would you say the best remedy would be to allow frame_time to be an array so that different timings can be applied to each image in the sequence? Or is there another way to do this that could be more straightforward? 
many thanks in advance,
Spiro

**Comment by jodeleeuw** [comment]
The easiest way to set this up is to just have two consecutive trials from the single-stim plugin. The first trial would having a timing_response parameter of 33 and timing_post_trial of 0.

However, 33ms is in the territory where standard JavaScript techniques will become unreliable. Monitor refresh rates need to be considered at such small intervals. JavaScript can sort of do this, with `requestAnimationFrame`, but this isn't supported in jsPsych yet.

**Comment by spiropan** [comment]
Thanks Josh, JSpsych looks like a very useful toolbox, thank you for
developing it! Any idea when that feature would be supported in JSPSych?
Thanks, Spiro

**Comment by jodeleeuw** [comment]
It's on the list of issues: #75 

But it's a complicated change and is not the highest priority at the moment, so I don't have an estimated date.

**Comment by nikbpetrov** [comment]
The `requestionAnimiationFrame` is supported with the [psychophysics plugin](https://jspsychophysics.hes.kyushu-u.ac.jp/). The author [suggests](https://jspsychophysics.hes.kyushu-u.ac.jp/faq/#3-how-many-frames-of-stimuli-will-the-plugin-present-if-i-specify-a-presentation-time-of-40-ms) that specifying the number of frames, instead of milliseconds, optimises performance, especially in a masking task.

I am working on a masking task now so if anyone has achieved this already, I'd appreciate tips. My biggest concern is how on earth can I be certain that the stimulus was presented for exactly that time I want it? The psychophysics plugin gives me the average frame rate in the data, but if I have 1000 frames and I have presented a stimulus for only a single frame, why do I have to rely on the average frame rate as an estimate for the presentation time of the stimulus as opposed to having access to the exact timing? And even if I record the time myself (by using, say, `performance.now()` before stimulus presentation and after, and then taking the difference) is that a reliable estimate? How do I go about verifying that?

In any case, I hope this helps future users.
